src/phoenix/experimental/evals/models/openai.py

```diff
@@ -151,11 +241,82 @@ def _build_messages(
     def _verbose_generation_info(self) -> str:
         return f"OpenAI invocation parameters: {self.public_invocation_params}"
 
-    def _generate(self, prompt: str, **kwargs: Dict[str, Any]) -> str:
+    def generate(
+        self, prompts: List[str], instruction: Optional[str] = None, num_consumers: int = 20
+    ) -> List[str]:
+        try:
+            asyncio.get_running_loop()
+            nest_asyncio.apply()
+        except RuntimeError:
+            pass
```

Comment:
Recap of a conversation @anticorrelator and I had. `nest_asyncio.apply` is used here so that `asyncio.run` can work in Jupyter notebooks, which already have running event loops. It has the side-effect of globally patching the internals of `asyncio`. Ideally, we'd avoid that.

A few options:

- We could run the `asyncio` code in a separate thread. This would require making `LimitStore` thread-safe and introduces complexity.
- We could leave it up to the client to deal with this issue. That's not very ergonomic.
- We could make this more explicit in `llm_classify` and `llm_generate`, so that there is a flag such as `nest_asyncio` defaulting to true so that people have some way of knowing that this is happening.
- We detect and use the previously running event loop, which has the potential to cause deadlocks.